### PR TITLE
feat(pytorch_poc): implement ReRoPE

### PR DESCRIPTION
**Pull request by tpoisonooo**
## Motivation

llama2 support long context.

## Modification

* add rerope feature for llama2 
* add passkey_retrieval testcase
* fix `rms_norm` missing mask

## Precision

tested on passkey_retrieval

```bash
input_token_length     precision
8128                           1.0
8947                           1.0
10040                         1.0
11132                          1.0
11951                         1.0
13043                         1.0
14136                         1.0
14955                         1.0
OOM
```

### CODE REVIEW

File: lmdeploy/pytorch_poc/engine/engine.py [hunk 1]
```diff
@@ -456,14 +462,14 @@ class Engine:
...
-        trust_remote_code=True,
-    ) -> None:
+    def __init__(self,
+                 model_path: str,
+                 scheduler_config: SchedulerConfig = None,
+                 cache_config: CacheConfig = None,
+                 tp: int = 1,
+                 trust_remote_code=True,
+                 json_config_file: str = 'config.json',
+                 max_session_len=2048) -> None:
```
Comment (grimoire) [inline lmdeploy/pytorch_poc/engine/engine.py:472]: max_session_len can be passed by scheduler_config

File: lmdeploy/pytorch_poc/kernels/rerope_attention.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        l_i = l_i * alpha + tl.sum(p, 1)
+        m_i = m_i_new
+        # update pointers
+        K1_block_ptr = tl.advance(K1_block_ptr, (0, BLOCK_N))
+        K2_block_ptr = tl.advance(K2_block_ptr, (0, BLOCK_N))
+        V_block_ptr = tl.advance(V_block_ptr, (BLOCK_N, 0))
+    # write back l and m
+    acc = acc / l_i[:, None]
+    l_ptrs = L + off_hz * N_CTX + offs_m
+    tl.store(l_ptrs, m_i + tl.math.log2(l_i))
```
Comment (grimoire) [inline lmdeploy/pytorch_poc/kernels/rerope_attention.py:143]: Do we have to store L?

File: lmdeploy/pytorch_poc/kernels/rerope_attention.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    m_i = tl.zeros([BLOCK_M], dtype=tl.float32) - float('inf')
+    l_i = tl.zeros([BLOCK_M], dtype=tl.float32)
+    acc = tl.zeros([BLOCK_M, BLOCK_DMODEL], dtype=tl.float32)
+    # scale sm_scale by log_2(e) and use
+    # 2^x instead of exp in the loop because CSE and LICM
+    # don't work as expected with `exp` in the loop
+    qk_scale = sm_scale * 1.44269504
+    # load q: it will stay in SRAM throughout
+    q1 = tl.load(Q1_block_ptr, boundary_check=(0, 1))
+    q1 = (q1 * qk_scale).to(tl.float16)
```
Comment (grimoire) [inline lmdeploy/pytorch_poc/kernels/rerope_attention.py:92]: Can we cast to q1.dtype so we can support both fp16, bf16 ... etc.

File: lmdeploy/pytorch_poc/kernels/rms_norm.py [hunk 4]
```diff
@@ -6,42 +6,94 @@
 
 
 @triton.jit
-def rms_norm_kernel(X, WEIGHT, OUT, eps, BLOCK_N: tl.constexpr):
+def rms_norm_kernel(input, weight, output, input_row_stride, n_cols, eps,
```
Comment (grimoire) [inline lmdeploy/pytorch_poc/kernels/rms_norm.py:9]: n_cols is redundant of N_COLS

File: lmdeploy/pytorch_poc/models/functional.py [hunk 5]
```diff
@@ -2,17 +2,43 @@
...
 
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., :x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2:]
+    return torch.cat((-x2, x1), dim=-1)
+
+
+def repeat_kv(hidden_states: torch.Tensor, n_rep: int) -> torch.Tensor:
+    """This is the equivalent of torch.repeat_interleave(x, dim=1,
```
Comment (grimoire) [inline lmdeploy/pytorch_poc/models/functional.py:27]: So what is the difference between repeat_kv and repeat_interleave?

File: lmdeploy/pytorch_poc/models/llama.py [hunk 6]
```diff
@@ -64,33 +66,130 @@ def _contiguous_batching_forward_impl(
...
+
+                position_ids = (position_ids[:, -1] -
+                                position_ids).clip(max=window)
+                _, key_states = apply_rotary_pos_emb_rerope(
+                    None, key_states, cos, -sin, position_ids)
+
+                if self.num_key_value_groups > 1:
+                    key_states = repeat_kv(key_states,
+                                           self.num_key_value_groups)
+                    value_states = repeat_kv(value_states,
```
Comment (grimoire) [inline lmdeploy/pytorch_poc/models/llama.py:144]: value_states has not been returned. Is this repeat_kv necessary?

**Review by wangruohui — APPROVED** [review]

File: lmdeploy/pytorch_poc/kernels/rms_norm.py (hunk 4 above)
Comment (tpoisonooo) [reply lmdeploy/pytorch_poc/kernels/rms_norm.py:9]: `type(n_cols)` is tensor while `type(N_COLS)` is int, not same one.

File: lmdeploy/pytorch_poc/models/functional.py (hunk 5 above)
Comment (tpoisonooo) [reply lmdeploy/pytorch_poc/models/functional.py:27]: "This is the equivalent of torch.repeat_interleave"

File: lmdeploy/pytorch_poc/engine/engine.py (hunk 1 above)
Comment (tpoisonooo) [reply lmdeploy/pytorch_poc/engine/engine.py]: fixed.

File: lmdeploy/pytorch_poc/engine/engine.py [hunk 7]
```diff
@@ -834,6 +849,7 @@ def step(self, return_logits=False):
         next_token_ids = []
         for msg, logit, param in zip(running, split_logits, sampling_params):
             input_ids = torch.tensor(msg.token_ids)
+            input_ids = msg.token_ids.clone().detach()
```
[the remark this answers is not in the document]
Comment (tpoisonooo) [reply lmdeploy/pytorch_poc/engine/engine.py]: fixed.

File: lmdeploy/pytorch_poc/kernels/rerope_attention.py (hunk 2 above)
Comment (tpoisonooo) [reply lmdeploy/pytorch_poc/kernels/rerope_attention.py]: L is for debugging,  fixed.

**Review by grimoire — APPROVED** [review]
